storage/innobase/trx/trx0rseg.cc

```diff
@@ -571,12 +616,41 @@ trx_rseg_array_init()
 				ut_ad(!trx_sys.rseg_array[rseg_id]);
 				trx_sys.rseg_array[rseg_id] = rseg;
 				trx_rseg_mem_restore(rseg, max_trx_id, &mtr);
+#ifdef WITH_WSREP
+				if (!wsrep_sys_xid.is_null() &&
+				    !wsrep_sys_xid.eq(&trx_sys.recovered_wsrep_xid)) {
+					wsrep_xid_in_rseg_found = true;
+#ifdef UNIV_DEBUG
+					if (!memcmp(wsrep_xid_uuid(&wsrep_sys_xid),
+						    wsrep_xid_uuid(&trx_sys.recovered_wsrep_xid),
+						    sizeof wsrep_uuid)) {
+						ut_ad(wsrep_xid_seqno(&wsrep_sys_xid) <=
+						      wsrep_xid_seqno(&trx_sys.recovered_wsrep_xid));
+					}
+#endif
+				}
+#endif
 			}
 		}
 
 		mtr.commit();
 	}
 
+#ifdef WITH_WSREP
+	if (!wsrep_sys_xid.is_null()) {
+		/* Upgrade from a version prior to 10.3.5,
+		where WSREP XID was stored in sys header.
+		If no rollback segment has a WSREP XID set,
+		we must copy the XID found in sys header to
+		rollback segments. Finally, clear the WSREP
+		XID persisted in system header. */
+		if (!wsrep_xid_in_rseg_found) {
+			trx_rseg_update_wsrep_checkpoint(&wsrep_sys_xid);
+		}
+		trx_rseg_clear_wsrep_sys_header_checkpoint();
```

Comment:
The comment (as well as the commit message) should be more specific and say "TRX_SYS page" instead of "sys header".
I think that the section of the TRX_SYS page should be cleared only when needed, and as part of the same atomic mini-transaction. That is, the write should be invoked from `trx_rseg_update_wsrep_checkpoint()`.